### PR TITLE
ENH, BLD: Add windows arm64 support

### PR DESCRIPTION
**Pull request by ader1990**
Added support for building numpy for Windows ARM64:

  * fix philox generator to use correct arm64 intrinsic __umulh
  * fix cpu recognition

Test results:

```bash
python.exe -c "import numpy; numpy.test(label='fast',verbose=3)

>> 11404 passed, 705 skipped, 1205 deselected, 19 failed in 497.70s (0:08:17)
```

You need a tweaked version of the msvc compiler for Python, which you can download from here: https://github.com/ader1990/CPython-Windows-ARM64/releases
Source code for Python 3.8.5: https://github.com/ader1990/cpython/tree/v3.8.5-win-arm64 (basically is just a hack to recognise the emulated build tools for ARM64). You can rebase the commit that fixes the build on recent Python versions too: https://github.com/ader1990/cpython/commit/b8c59c9b96a7ad11094224b5631aae3b89323f7a

This fix is an alternative solution for https://github.com/numpy/numpy/issues/17747

### CODE REVIEW

File: numpy/random/src/philox/philox.h [hunk 1]
```diff
@@ -32,6 +32,17 @@ static NPY_INLINE uint64_t mulhilo64(uint64_t a, uint64_t b, uint64_t *hip) {
...
+#elif defined(_M_X64) || defined(_M_ARM64)
+#include <intrin.h>
+#pragma intrinsic(__umulh)
+static NPY_INLINE uint64_t mulhilo64(uint64_t a, uint64_t b, uint64_t *hip) {
+  uint64_t p0;
+  uint64_t p1;
+  p0 = a * b;
+  p1 = __umulh(a, b);
+  *hip = p1;
+  return p0;
```
Comment (eric-wieser) [inline numpy/random/src/philox/philox.h:44]: nit:
```suggestion
  uint64_t p0 = a * b;
  uint64_t p1 = __umulh(a, b);
  *hip = p1;
  return p0;
```
or if you don't think it harms clarity
```suggestion
  *hip = __umulh(a, b);
  return a * b;
```

Comment (ader1990) [reply numpy/random/src/philox/philox.h:44]: I would like to keep the p0, p1 namings, as the compiler will make the correct optimizations and the same jargon is used in the documentation I found available.

**Comment by eric-wieser** [comment]
> This fix is an alternative solution for #17747

I'm not sure I understand the comparison. Are you suggesting that a different fix to #17747 would make this unecessary?

**Comment by ader1990** [comment]
> > This fix is an alternative solution for #17747
> 
> I'm not sure I understand the comparison. Are you suggesting that a different fix to #17747 would make this unecessary?

Well, in the sense that there were multiple solutions presented on that issue (like future x64 emulation). Of course, I think the native build solution would be the best.

File: numpy/core/include/numpy/npy_cpu.h [hunk 2]
```diff
@@ -39,6 +39,12 @@
      * _M_AMD64 defined by MS compiler
      */
     #define NPY_CPU_AMD64
+#elif defined(_M_ARM64)
+    /*
+     * _M_ARM64 is defined by VS compiler >= 2017 to build native
+     * ARM64 packages on Windows
+     */
+    #define NPY_CPU_ARMEL_AARCH64
```
Comment (mattip) [inline numpy/core/include/numpy/npy_cpu.h:47]: why `ARMEL_AARCH64` ? and not `MSVC_AARCH64`?

File: numpy/random/src/philox/philox.h [hunk 3]
```diff
@@ -32,6 +32,15 @@ static NPY_INLINE uint64_t mulhilo64(uint64_t a, uint64_t b, uint64_t *hip) {
   *hip = product >> 64;
   return (uint64_t)product;
 }
+#elif defined(_M_X64) || defined(_M_ARM64)
+#include <intrin.h>
+#pragma intrinsic(__umulh)
+static NPY_INLINE uint64_t mulhilo64(uint64_t a, uint64_t b, uint64_t *hip) {
+  uint64_t p0 = a * b;
+  uint64_t p1 = __umulh(a, b);
```
Comment (mattip) [inline numpy/random/src/philox/philox.h:40]: Is this an intrnisic newly defined for arm64 + MSVC? Is there a parallel intrinsic for arm64 + gcc?

File: numpy/core/include/numpy/npy_cpu.h (hunk 2 above)
Comment (ader1990) [reply numpy/core/include/numpy/npy_cpu.h:47]: As this is for endianess check, the closest was armel_aarch64.

File: numpy/random/src/philox/philox.h (hunk 3 above)
Comment (ader1990) [reply numpy/random/src/philox/philox.h:40]: Here is the general definition: https://docs.microsoft.com/en-us/cpp/intrinsics/umulh?view=msvc-160
And it is present as an arm64 intrinsic: https://docs.microsoft.com/en-us/cpp/intrinsics/arm64-intrinsics?view=msvc-160

I do not know if there is one for arm+gcc.

**Comment by zooba** [comment]
FYI, there are "unofficial official" releases of CPython for ARM64 at https://dev.azure.com/Python/cpython/_build?definitionId=21 (browse to a build result, then "Artifacts" and you probably want "layout_full_arm64").

When numpy is ready to claim support for ARM64, I'll be happy to make them official CPython releases. But I don't want you guys to feel like you've been forced into it, which is why I'm holding off for now (and simultaneously trying to get some CI hardware for you... but that's a bit 😫...)

**Comment by ader1990** [comment]
> FYI, there are "unofficial official" releases of CPython for ARM64 at https://dev.azure.com/Python/cpython/_build?definitionId=21 (browse to a build result, then "Artifacts" and you probably want "layout_full_arm64").
> 
> When numpy is ready to claim support for ARM64, I'll be happy to make them official CPython releases. But I don't want you guys to feel like you've been forced into it, which is why I'm holding off for now (and simultaneously trying to get some CI hardware for you... but that's a bit 😫...)

Unfortunately, I do not have access to those artifacts. The builds I created using github pipelines should be practically the same with those from that artifact store. Example of build produced using github actions for tag v3.9.0: https://github.com/ader1990/cpython-builder/actions/runs/377798403 . I have tested that build with this numpy patch and it works perfectly.

**Comment by zooba** [comment]
You can't even download them from the "..." at the right? That ought to work without signing in...

Either way, the main difference should just be that these ones are signed (and will match every release made from python.org).

**Comment by charris** [comment]
@ader1990 Ping.

@zooba Any update on progress getting the ARM infrastructure in Azure?

**Comment by ader1990** [comment]
> @ader1990 Ping.
> 
> @zooba Any update on progress getting the ARM infrastructure in Azure?

Unfortunately, I cannot access the Azure infra, but the code looks great and it works great IMHO.

**Comment by zooba** [comment]
> Any update on progress getting the ARM infrastructure in Azure?

Let me be careful here... there is _progress_, but there is no _update_... yeah I think I can say that much. :)

File: numpy/random/src/philox/philox.h (hunk 3 above)
Comment (bashtage) [reply numpy/random/src/philox/philox.h:40]: Is this better on AMD64 than using `_umul128` directly?

**Comment by charris** [comment]
@ader1990  I'm closing this in favor of #19513 which has been merged. Thanks for the initial work.

**Pull request closed without merging**